### ISSUE TITLE
Generalizing to other force field generators?

### ISSUE THREAD

**Issue by jchodera** [chapincavender/proteinbenchmark#8]
We'd love to benchmark `espaloma` with this benchmark framework. To do this, we'd need to recognize espaloma force field names as input, and generate an `Interchange` object (or an OpenMM `System`) as output. 

Is there a single place in the code where this should happen? If not, could the various places this happens be coalesced into one place to make it easier to extend?

If it's already structured this way, which part of the code would need to be extended to add support for `espaloma` parameterization of complete OpenFF `Topology` objects?

**Comment by chapincavender** [comment]
@jchodera Thinking about this more carefully, I think this project needs a small refactor to be able to support other force field generators.

After building initial coordinates for the solute, the code to produce a solvated, parametrized system is contained in [this function](https://github.com/openforcefield/proteinbenchmark/blob/ffb8bec1038690adde82b39518bd948c24c5725f/proteinbenchmark/system_setup.py#L268). The workflow looks like this:
- Construct a force field object from a path specified in the dictionary [here](https://github.com/openforcefield/proteinbenchmark/blob/ffb8bec1038690adde82b39518bd948c24c5725f/proteinbenchmark/force_fields.py#L10)
- Add water and ions using OpenMM Modeller
- Assign parameters and export solvated system to format expected for desired MD simulation engine, currently OpenMM or GROMACS

For the first and third steps, OpenMM force fields and SMIRNOFF force fields are handled separately by checking the file extension. This could be extended to handle other generators by introducing a key `force_field_type` in the force field dictionary to specify how to handle the first and third steps in the workflow.

For the second step, the force field object must be able to be passed to Modeller's `addSolvent()` method, i.e. it must have a method `createSystem()` that takes an OpenMM Topology as input and creates a parametrized OpenMM System as output. This is natively supported by OpenMM force fields. For SMIRNOFF force fields, I created a [dummy wrapper class](https://github.com/openforcefield/proteinbenchmark/blob/ffb8bec1038690adde82b39518bd948c24c5725f/proteinbenchmark/system_setup.py#L15) to provide the `createSystem()` method. You may have to do something similar for Espaloma. I think this is a clunky solution, so if you have ideas for alternatives to allow non-OpenMM force fields to interface with Modeller I'd be happy to implement something else.

**Comment by jchodera** [comment]
Thanks for the thoughtful analysis, @chapincavender!

Do you think it would make sense to separate the system setup part (addition of water and ions) from the parameterization step? 

The system setup currently just required *some* reasonable force field that won't throw an error or generate geometries that could not be minimized. We could use any reasonable general force field for this---potentially even some form of the dummy force field OpenMM includes. (I am also hoping we can support something extremely generic like UFF at some point.) Later, we expect there might be better tools for system setup available.

For the parameter assignment step, we just need some reasonable API for accepting an OpenFF Topology and producing an Interchange object with parameters. We could then easily replace that with anything (espaloma, other models, etc). 

The alternative approach you suggest has the difficulty that we ultimately have to shoehorn new force field.approaches into the OpenMM ForceField infrastructure that we are trying to move away from because it is inherently residue template based and does not carry true chemical topology information. We could continue to try to do this, but we are going to continue to invest a lot of effort in a dead end system until we break free from it.

**Comment by chapincavender** [comment]
I agree that separating the solvation and parametrization steps is the best solution to this problem. Your proposal will make the system setup workflow much more extensible.

Since this repo currently focuses on proteins, I think that it makes sense to use the OpenMM implementation of Amber ff14SB as a reference force field to build initial coordinates for the solvated system using Modeller.

I'll implement this idea and tag you when it's done. Then someone in your group can extend the parametrization section to take an espaloma generator.

**Comment by mrshirts** [comment]
@jchodera do we have personnel and a timeframe for implementing your proposal, or does that need to be assigned?

**Comment by mrshirts** [comment]
I'm now rereading and seeing @chapincavender is tagged as implementing the idea.  Does this still make sense given timelines (i.e. we would like to start running these calculations with espaloma in early September to make sure we have time to report them for the NIH grant).
